staging/src/k8s.io/apiserver/pkg/storage/selection_predicate.go

```diff
@@ -128,3 +129,21 @@ func (s *SelectionPredicate) MatchesSingle() (string, bool) {
 func (s *SelectionPredicate) Empty() bool {
 	return s.Label.Empty() && s.Field.Empty()
 }
+
+// For any index defined by IndexFields, if a matcher can match only (a subset)
+// of objects that return <value> for a given index, a pair (<index name>, <value>)
+// wil be returned.
+func (s *SelectionPredicate) MatcherIndex() []MatchValue {
```

Comment:
I've made it all the way down here and I need some help being able to make some sense of this coming in cold.  I can't believe I'm going to ask for comments, but I need some comments on the struct to make decisions about this method.

1. Conflating fields and labels in a single index will make it nearly impossible to have this generally configurable for things like CRDs in the future
2. Based on the names in this file, I expected that SelectionPredicate.Label and SelectionPrediate.Field would be based on user input from the request.  If I'm mistaken, can you add doc explaining where they come from?  `generic/registry/store.go|List` as a for instance.
3. I think this indicates a need to create an e2e test (not a unit) for this feature that create overlapping field and label selectors that should select different, non-overlapping sets and provides proof that they are returned correctly.

Reply:
> Conflating fields and labels in a single index will make it nearly impossible to have this generally configurable for things like CRDs in the future

I'm not sure I'm following this one - I agree we have a naming problem (we should somehow distinguish them, but sth like "field:<field-name>" and "label:<label-name>" or sth like that should do the job). 
But I don't understand how CRDs make things worse for that.

> I think this indicates a need to create an e2e test (not a unit) for this feature that create overlapping field and label selectors that should select different, non-overlapping sets and provides proof that they are returned correctly.

At the high level I agree. But e2e test would be hard, given that there is no way to create indices other than programatically. I'm not entirely sure how to make non-unit test for it.

Reply:
It needs to be impossible for fields and labels to collide. We should change this code until we are confident that they don't and it's verifyable with a unit test.